dash/include/dash/algorithm/Sort.h

```diff
@@ -830,24 +834,40 @@ void sort(
   auto const lmax = (n_l_elem > 0) ? sortable_hash(*(lend - 1))
                                    : std::numeric_limits<mapped_type>::min();
 
-  dash::Shared<dash::Atomic<mapped_type>> g_min(dash::team_unit_t{0}, team);
-  dash::Shared<dash::Atomic<mapped_type>> g_max(dash::team_unit_t{0}, team);
-
-  g_min.get().op(dash::min<mapped_type>(), lmin);
-  g_max.get().op(dash::max<mapped_type>(), lmax);
+  dash::team_unit_t const                 owner{0};
+  dash::Shared<dash::Atomic<mapped_type>> g_min(lmin, owner, team);
+  dash::Shared<dash::Atomic<mapped_type>> g_max(lmax, owner, team);
 
   using array_t = dash::Array<std::size_t>;
 
   std::size_t gsize = nunits * NLT_NLE_BLOCK * 2;
 
+  // implicit barrier...
   array_t g_nlt_nle(gsize, dash::BLOCKCYCLIC(NLT_NLE_BLOCK), team);
 
+  if (myid != owner) {
+    // the other units apply min / max reductions
+    g_min.get().op(dash::min<mapped_type>(), lmin);
```

Comment:
Are we sure this is actually more efficient than using `allreduce`?  `allreduce` should be somewhere in *O(logn)* (see Rabenseifner's algorithm for example) and is usually a well optimized MPI functionality while what we're doing here is *O(n)*.

Reply:
> I would argue that the O(logn) vs O(n) only applies in the very large scale (> 1000 nodes) which is also what previous benchmark results confirm.

There might be a higher constant overhead involved for smaller numbers of units, although I would expect it to become significant way earlier than 1000 nodes. The problem with atomics is that only few implementations really support them in hardware. My measurements indicate that only Open MPI on IB networks actually does asynchronous progress. Cray is still struggling with it (according to one of their developers) and I haven't seen positive results from any other MPICH-based implementation (namely Intel MPI). For atomics implemented in software, this results in linear scaling of two-sided messages (to send the request for the atomic operation and receive the results) and work on the target. I'd be interested to see the sweet-spot in some measurements though.

> On the other hand, the one-sided counterpart causes less synchronization, i.e. allreduce implies synchronization in a barrier-like manner. So it is more efficient for a more or less low node count while the collective two-sided counterparts come into play with a higher node count.

Looking at the code, the benefit of one-sided communication seems completely voided by the immediately preceding and succeeding creations of global arrays, which are synchronizing operations. The processes are in lock-step anyway. Plus, the setup of a `dash::Shared` involves a collective operation as well...

Having said that, this should not impact the fixing of the bug this PR addresses. However, I think we need to carefully assess where collective operations make sense (and I think there is definitely room for improvements in other DASH algorithms as well).

Reply:
It makes sense what you are saying and I definitely agree with that. However, since we can only speculate, lets do some benchmarks. I can write a small benchmark over the next weekend and run that for various unit / node counts, both shared and distributed memory.
I claim that achievable performance depends on two parameters. First, the number of units and nodes. The more cores involved the greater the benefit of two-sided MPI. The other parameter what may be interesting is the density. As an example `dash::sort` may be called only on a specific range of the container. Additionally, we may have "holes" in our range in such a way that we have units which have an empty local range. This may be interesting for sparse data structures. Depending on the density, i.e. the ration of contributing units (non-empty local portion which is part of the range to be sorted) and the number of units in my team, one-sided communication may be more efficient. Let us say the ratio is 50-60% and we have the advantage in DASH that we know this at runtime, we may switch between collective or one-sided communication.